### ISSUE TITLE
Reuse query engine for multiple `PrismaClient` instances

### ISSUE THREAD

**Issue by timsuchanek** [prisma/prisma-client-js#730]
## Problem
One of the most tricky parts of using Prisma Client is database connection handling.
While modern databases like Fauna DB or serverless Aurora don't require a long-running connection anymore and therefore no long-running process that keeps the connection up, the databases currently supported by Prisma (in particular MySQL and Postgres) have the concept of long-running connections, which are expensive to establish and once established should be reused.

In order to ensure optimal performance with these databases, the query engine, which powers Prisma Client, manages a connection pool. The default size of that connection pool is being determined by the number of cores times 2, as it turned out in internal performance tests, that that is the optimal number for the best performance.

All this works well, as long as there is only one Prisma Client and therefore query engine instance.
And usually, as you have your application and database under control (should at least), you can ensure, that there is just one Prisma Client instance, which holds the query engine with the connection pool.

However, as soon, as you have multiple Prisma Client instances running, this can get really slow. It can even be a memory leak if you instantiate a new Prisma Client per incoming API request.

This situation, that too many Prisma Clients have been instantiated has occurred numerous times in the community, see https://github.com/prisma/prisma/issues/1983 https://github.com/prisma/prisma-client-js/issues/632 https://github.com/prisma/prisma-client-js/issues/723 and https://github.com/blitz-js/blitz/blob/canary/examples/store/db/index.ts

In other words: You can easily shoot yourself in the foot because you need to manage state and managing state is always nasty.

## How could Prisma Client help here?
One possible solution: Instead of instantiating a new query engine process per `PrismaClient` instance, we would have a "named singleton" pattern.
That means, that for a particular combination of args, we **reuse** the query engine instance and make that accessible under a name, e.g. "QE1". This way it doesn't matter, how often you instantiate PrismaClient, all instances will reuse the same query engine and therefore the same connection pool, with all the connections already warmed up.
This could also help us in the situation of clusters: https://github.com/prisma/prisma-client-js/issues/632  we could enable this singleton pattern across the whole cluster out of the box so that the node cluster can share the same connection pool. Building something like that as a user would be a lot of overhead and a non-trivial task to do.

## What are disadvantages of this approach?
Usually, a singleton is considered an antipattern. In this case it could be considered "ok" though, as we're talking about resources that can be shared (and maybe should be shared) across instances.

Another big problem here is: You might indeed have a serious memory leak in your application, which would now not be visible anymore, as it "just works".
One could see this approach as "wrong", as you can always say: "Just instantiate  one Query Engine!".

## Closing thoughts
We have this problem because the query engine is attached as a sidecar to the Node.js process. If we would run the query engine as a separate process as you do with a database, we would not have this problem of "unintentionally spawning too many query engines". However, the ergonomy and developer experience of Prisma Client would then greatly suffer.

**Comment by Sytten** [comment]
A couple of points:
- You would also not have this problem with direct bindings (not just a separate process)
- Sometimes it's tricky how the `PrismaClient` is created, so I do support the use of a singleton (I was against before having some issues myself)
- I believe if an attempt is made to "repawn" a new client, a warning should be displayed to the user by default so he is informed that he tried to spawn a new process but the client catched it (but make it so we can disable that with a flag if this is the intended behaviour)

**Comment by timsuchanek** [comment]
> You would also not have this problem with direct bindings (not just a separate process)
That highly depends on the process model we would choose with the direct bindings.

> I believe if an attempt is made to "repawn" a new client, a warning should be displayed to the user by default so he is informed that he tried to spawn a new process but the client caught it (but make it so we can disable that with a flag if this is the intended behaviour)

Yes, we could additionally add a warning for that case.

**Comment by matthewmueller** [comment]
> This could also help us in the situation of clusters: #632 we could enable this singleton pattern across the whole cluster out of the box so that the node cluster can share the same connection pool.

I don't think this is quite right. As far as I understand node cluster, it's a master process + N worker processes. 

The singleton will occupy the master processes' memory space, but not be accessible to the worker processes.

I agree that a default singleton is probably the most convenient out of the box, plus it requires no API changes. I'd like to suggest we separate these out cleanly. 

The use case I have in mind is a single Prisma Engine on a separate process, probably locally but it could be over a network. I think it's possible to support our current setup and this setup quite easily with the right abstraction.

Some psuedo-code:

```ts
import { PrismaClient, QueryEngine } from "@prisma/client"

// the default, works the same way as today
// but uses a QueryEngine singleton under the hood
const client = new PrismaClient()

// alternatively, spawn a query engine yourself
const qe = new QueryEngine(options)
// and pass the query engine process into the prisma client
const client = new PrismaClient({ process: qe })

// alternatively, to connect to a query engine on a separate process
const qe = new QueryEngine(options)
qe.listen(port)

// connect to that query engine with some known URL.
const client = new PrismaClient({ url: queryEngineURL })
```

**Comment by timsuchanek** [comment]
> I don't think this is quite right. As far as I understand node cluster, it's a master process + N worker processes. The singleton will occupy the master processes' memory space, but not be accessible to the worker processes.

Exactly, the singleton would run in the master, and in the workers, we would make it available via [ipc communication](https://gist.github.com/jpoehls/2232358). Users will just import from `@prisma/client` but in our code, we check with `require('cluster').isWorker` if we need to talk to the master to do queries. This is what I meant with "Building something like that as a user would be a lot of overhead and a non-trivial task to do.". We can do this as a library once to solve this for all users.

While I think that your API proposal gives more flexibility and could for sure be something that advanced users might be interested in, the point of this issue was, that even the current API can be a foot gun and especially for more beginner users to hard to manage.
If we now even go one level down in abstraction, we would introduce another level of complexity for the user to manage.

I like that idea, but I would see that as an advanced API.

**Comment by matthewmueller** [comment]
> While I think that your API proposal gives more flexibility and could for sure be something that advanced users might be interested in, the point of this issue was, that even the current API can be a foot gun and especially for more beginner users to hard to manage.

The singleton pattern can protect against this case. I'm 👍  to that.

> Users will just import from @prisma/client but in our code, we check with require('cluster').isWorker if we need to talk to the master to do queries

This feels like an overreach. We're now managing 3 processes, their lifecycles, their IPC communication, error states behind: `const client = new PrismaClient()`. Certainly possible, it will just lead to more configuration being moved into `new PrismaClient(config)`.

Personally, I'd rather we suggest using the advanced API once you're reaching for node's cluster capabilities. This will allow us to more gracefully move to Prisma engine working across machines.
